### PR TITLE
Refactor ProgramsCache into ChainPrograms

### PR DESCRIPTION
**Pull request by pattyshack**
The main difference between the two is in how we create BlockPrograms for block execution.  The HasChanges optimization no longer makes sense since cloning BlockPrograms now deep copied the map in order to support fine grain eviction.

### CODE REVIEW

File: fvm/programs/chain_programs.go [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+
+const DefaultProgramsCacheSize = 1000
+
+// ChainPrograms is a cache of BlockPrograms databases used for speeding up
+// cadence execution.
+//
+// Since programs are derived from external source, the BlockPrograms databases
+// need not be durable and can be recreated on the fly.
+type ChainPrograms struct {
+	mutex sync.Mutex
```
Comment (Tonix517) [inline fvm/programs/chain_programs.go:20]: how about `RWMutex`?

File: fvm/programs/chain_programs.go [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+	if ok {
+		return entry.(*Programs)
+	}
+
+	return nil
+}
+
+func (chain *ChainPrograms) NewBlockProgramsForScript(
+	currentBlockId flow.Identifier,
+	parentBlockId flow.Identifier,
```
Comment (Tonix517) [inline fvm/programs/chain_programs.go:96]: nit: i assume the invariant here is that `parentBlockId` should be actually the id of `currentBlockId` 's parent node. But if we expose these params, there's a slight chance that caller may mistakenly call it with this invariant broken. not blocking to me, just sth to think about.

File: fvm/programs/chain_programs.go (hunk 1 above)
Comment (pattyshack) [reply fvm/programs/chain_programs.go:20]: It's not safe to use RWMutex in this case since lru will modify the data structure on Get

Comment (Tonix517) [reply fvm/programs/chain_programs.go:20]: nice catch.

**Review by Tonix517 — APPROVED** [review]

File: fvm/programs/chain_programs.go [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+) *Programs {
+	chain.mutex.Lock()
+	defer chain.mutex.Unlock()
+
+	currentEntry, ok := chain.lru.Get(currentBlockId)
+	if ok {
+		return currentEntry.(*Programs)
+	}
+
```
Comment (janezpodhostnik) [inline fvm/programs/chain_programs.go:63]: this block could be a method `(chain *ChainPrograms) get(blockId flow.Identifier) *Programs`.

**Review by janezpodhostnik — APPROVED** [review]
LGTM.

1 minor comment.

File: fvm/programs/chain_programs.go (hunk 3 above)
Comment (pattyshack) [reply fvm/programs/chain_programs.go:63]: my bad here.  we can just use Get here directly.  I've made some minor changes and forgot to clean this up

**Comment by pattyshack** [comment]
bors merge

**Pull request merged**